HeapSort<指摘対応完了06/12>

a3095194 opened the pull request:
配列を自由に入力可能にしました。
その為、要素数が自動で増やせるArrayListクラスを使用しました。

a3095194 committed:
HeapSort<修正中>

配列を自由に入力可能にしました。
その為、要素数が自動で増やせるArrayListクラスを使用しました。

a3095194 commented:
calculateHeap()の中に
// ここを1にすると頂点がいつまで経っても変えられないが、0にすると子の場所が正確にとれない
とコメント入れてありますが、main()でダミーデータを仕込むことで先ほどより綺麗に動くようになりました。
まだ何点か直したいとこあるので、修正中のステータスとさせていただきました。

a3095194 committed:
HeapSort<修正中>

文字コードをANSI→UTF8に修正した。

a3095194 committed:
HeapSort<完了>

配列の出力をmain()より分割
コメント等もより分かり易く修正

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+		ArrayList<Integer> orgArray = new ArrayList<Integer>();
+		// 最初に入力されたデータのインデックスを[1]とするためダミーデータを格納
+		orgArray.add(999);
+		for (int i = 0, length = args.length; i < length; i++) {
```
@a3095194 
for-each文 使えないですかね？
https://github.com/a3095194/algorithm/pull/6/files

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+		arrayPrintln(orgArray);
+		System.out.print("並べ替え後：");
+		// 並べ替えの済んだ配列を格納
+		ArrayList<Integer> newArray = calculateHeap(orgArray);
```
[直さなくて良いです]
calculateHeap
計算ではなくソートな気がするので
sortingHeepとか?

newArrayも,
sortedHeepとかかな

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+	public static void main(String[] args) {
+		ArrayList<Integer> orgArray = new ArrayList<Integer>();
+		// 最初に入力されたデータのインデックスを[1]とするためダミーデータを格納
+		orgArray.add(999);
```
@a3095194 ソート要のダミーデータだと思うのでcalculateHeapの中でやってあげましょう。

kazuaking replied:
[直しても直さなくてもいいです]

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+	 * @param 配列
+	 */
+	public static void arrayPrintln(ArrayList<Integer> array) {
+		// ダミーデータを表示しないようにarray[1]から出力
```
@a3095194 ここでダミーデータを除外するよりは、calculateHeapでdeleteするような感じでcalculateHeap内で完結させるほうがよいでしょう。

kazuaking replied:
[直しても直さなくてもいいです]

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+				if (orgArray.get(leftChild) < orgArray.get(rightChild)) {
+					int temp = orgArray.get(leftChild);
+					orgArray.set(leftChild, orgArray.get(rightChild));
+					orgArray.set(rightChild, temp);
```
@a3095194 
[直しても直さなくてもいいです]

``` java
                    int temp = orgArray.get(leftChild);
                    orgArray.set(leftChild, orgArray.get(rightChild));
                    orgArray.set(rightChild, temp);
```

このような処理は毎回必要になっているのでメソッド化するのがよいでしょう
`replace(first, second)`
見たいなメソッドのイメージです。
int tempがスコープの外側でも重複して宣言されているので気持ち悪いという理由もあります。
(javaがこれを許してるのか? とちょっともやもや)

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+		// 並べ替えの済んだ配列を格納
+		ArrayList<Integer> newArray = calculateHeap(orgArray);
+		// 並べ替え後の配列を出力
+		arrayPrintln(newArray);
```
[ちなみに]
この時点でorgArrayとnewArrayは同じものですね。
それで、並べ替え前を処理実施前にprintしてるのかな。
これはJavaの仕様で引数が参照値を渡す仕様になってるからですね。

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+	 * @return ソート後の配列
+	 */
+	public static ArrayList<Integer> calculateHeap(ArrayList<Integer> orgArray) {
+		for (int i = orgArray.size() - 1; i >= 1; i--) {
```
https://github.com/a3095194/algorithm/pull/6/files#r66196999
に書いた理由から、引数にたいして更新をするのは良くないとされています。
新しいListのインスタンスを作成してあげましょう。

kazuaking replied:
これは直して欲しいかも

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,88 @@
+				System.out.println("数字以外が入力されました");
+			}
+			// 入力された数字をStringからintへ型変換後、ArrayListに詰める
+			orgArray.add(CastUtil.strToInt(args[i]));
```
値のチェックと、castの処理、確かに一度にやった方が効率的ですね。
ただ、この２つはやりたいことが異なるので、別の処理としてあげるのがよいでしょう。

kazuaking replied:
[直さなくてもいいです]

kazuaking commented:
みました。

a3095194 committed:
HeapSort<指摘対応完了>

ダミーデータの挿入と除去をソートする為のメソッド内で完結
型変換とList型配列への追加を別処理として変更
calculateHeap()の名前変更
メソッド引数ではなく、編集用配列を用意してしてそちらを書き換えるように変更
配列内データの入れ替えのみを行う共通クラスを作成

a3095194 commented:
@kazuaking 
ダミーデータを入れた位置に関してですが、自分は必要なデータたちが入ってしまう前にダミーデータを入れないと、[0]に入っているデータがダミーデータに上書きされてしまうものだと思っていたので、あの位置にしていました。

kazuaking commented:
@a3095194 はい。ちなみにちゃんと検証出来てませんがダミーデータを入れるというのは最後の手段で使う手で、
本社研修では突っ込まれる可能性があります（ソース見て判断出来なかったのでスルーしてしまいました。ごめんなさい。

a3095194 commented:
@kazuaking 
承知しました。何となく感じてはいましたが、奥の手みたいなものなんですね。
覚悟しておきます。

kazuaking commented:
まぁあまりいい方法じゃないってことですね。
それよりもcountとかをいい感じにしてダミーが要らなくなるソースが良いソースなんだという感じです。
これはこのままで良いです。

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,97 @@
+	 */
+	private static ArrayList<Integer> sortingHeap(ArrayList<Integer> orgArray) {
+		// 編集用のList型配列を作成
+		ArrayList<Integer> editingArray = orgArray;
```
@a3095194 これだと結局引数と戻り値は同じメモリのアドレスを指した状態は変らないですね。
あえて正解は出さないので調べてみてください。(研修で当たったらそこで調べてよいです)

kazuaking replied:
ここは頑張って直そう！

a3095194 replied:
すみません。
細かいやりとりはAdd a line noteでいいのですね。
覚えときます

a3095194 commented:
@kazuaking 
もしかしてclone()とかですかね。
これでは参照してるメモリ変わってないですね確かに。失礼しました。
帰宅したらちょっと調べてみます。

kazuaking commented:
Add a line noteで返信貰えると助かる。

kazuaking commented on Mondai/src/jp/ditgroup/common/Replace.java:
```diff
@@ -0,0 +1,23 @@
+	 *            大きいデータのインデックス
+	 * @return 入れ替え後の配列
+	 */
+	public static ArrayList<Integer> doReplace(ArrayList<Integer> array, int first, int second) {
```
@a3095194 
この場合引数のArrayListが変更されるのでvoidで良いかも(嘘)
ラッパークラスを作成するという方法もあるのですが、それはおいおい

@kazuaking 
このコメントの意味は理解できたのですが、(嘘)というのは最終的にどちらがベターということに落ち着くのでしょうか？
話の流れ的に返り値なしで作成した方が無難ってことで合ってますか？

kazuaking replied:
直さなくて良いです。

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,97 @@
+			// 入力された数字をStringからintへ型変換
+			int data = CastUtil.strToInt(args[i]);
+			// 変換した数字をList型配列に入れる
+			orgArray.add(i, data);
```
@a3095194 ごめんなさい
https://github.com/a3095194/algorithm/pull/6#discussion_r66197241
の対応だと思うけど、そういう意味じゃなかった。
ループを別にして、ということ。

``` java
orgArray.add(CastUtil.strToInt(args[i]));
```

というのは悪くないです（むしろすき)

kazuaking commented:
@a3095194 みました。
https://github.com/a3095194/algorithm/pull/6#discussion_r66398357
だけ対応をお願いします。

https://github.com/a3095194/algorithm/pull/6/files#r66399816
これは意図が伝わればそれで良いです。

kovayashi commented:
バリデーション内で配列に詰める処理を一緒にしたのは、
多分自分が無駄なfor文なくせって言ったために意識したのかなと思います。

ですが、伊藤さんが言うように目的が違う処理を一緒にするのは基本避けたほうが良いです。
無駄なfor文云々の指摘で、main内で出力させるという決まりがあればその限りではありません的なコメントをした記憶がありますが、目的（処理の単位）があってそれを実現するためならばこのままでいいですよ的な意味でした。

a3095194 committed:
HeapSort<指摘対応完了06/12>

入力チェックとキャスト後配列に詰めるfor文を分割
sortingHeap()内で新たな配列を用意してメソッド引数をコピー(ディープコピー)

kazuaking commented on Mondai/src/jp/ditgroup/algorithm/HeapSort.java:
```diff
@@ -0,0 +1,97 @@
+				System.out.println("数字以外が入力されました");
+			}
+		}
+		for (int j = 0, length = args.length; j < length; j++) {
```
[直さないでOK]
解説として、forで宣言される`int i`、`int j`はそのfor句の中だけで有効です。なので、ここの変数も`i`を使用しても大丈夫。
とはいえ見やすさとして別の変数名にするのはあり。
一応解説まで。

a3095194 replied:
forの中で宣言された変数はforの中のみ有効なのはしっかり覚えてます！
ご察しの通り、分けた方が見やすいかなと思い、int jとしました。マージして、帰宅したらソースツリーもpullしときます。

kazuaking commented:
goodです。特に直すべき指摘無し。
マージしてください。
